diff --git a/postgresql/libpq-version-generator/CMakeLists.txt b/postgresql/libpq-version-generator/CMakeLists.txt
index 125a5bfb7..b0253bbf1 100644
--- a/postgresql/libpq-version-generator/CMakeLists.txt
+++ b/postgresql/libpq-version-generator/CMakeLists.txt
@@ -1,4 +1,4 @@
 project(userver-impl-libpq-version-generator)
 
 add_executable(${PROJECT_NAME} "${CMAKE_CURRENT_SOURCE_DIR}/main.cpp")
-target_link_libraries(${PROJECT_NAME} PRIVATE userver-libpq)
+target_link_libraries(${PROJECT_NAME} PRIVATE userver-libpq ssl crypto pq)
diff --git a/postgresql/pq-extra/CMakeLists.txt b/postgresql/pq-extra/CMakeLists.txt
index b5bd233d9..8ce6af303 100644
--- a/postgresql/pq-extra/CMakeLists.txt
+++ b/postgresql/pq-extra/CMakeLists.txt
@@ -18,7 +18,7 @@ cmake_policy(SET CMP0067 NEW)
 try_compile(
     COMPILE_RESULT ${CMAKE_CURRENT_BINARY_DIR}/require_pq_extra
     SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/check_main.c ${PQ_EXTRA_SOURCES}
-    LINK_LIBRARIES PostgreSQLInternal
+    LINK_LIBRARIES PostgreSQLInternal ssl crypto pq
     OUTPUT_VARIABLE COMPILE_OUTPUT
 )
 
